### PR TITLE
Add support for esp32 board "Heltec wifikit 32 v2"

### PR DESCRIPTION
**Pull request by szsoftware**
Have succesfully running Circuitpython on this board.

### CODE REVIEW

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# Schematic:    https://resource.heltec.cn/download/WiFi_Kit_32/WIFI_Kit_32_Schematic_diagram_V2.pdf
+#               https://resource.heltec.cn/download/WiFi_Kit_32/WIFI_Kit_32_Schematic_diagram_V2.1.PDF
+#               (same as V2, but V2.1 is better readable)
+# Pinout:   https://resource.heltec.cn/download/WiFi_Kit_32/WIFI_Kit_32_pinoutDiagram_V2.pdf
+#           https://resource.heltec.cn/download/WiFi_Kit_32/WIFI_Kit_32_pinoutDiagram_V2.1.pdf
+#           (exactly same as V2 [md5sum of pdf is 9ee18bb5abe0a41fe98825cc756e8e3e])
+
+CIRCUITPY_CREATOR_ID = 0x148E173C
+# TODO: What value should it be?
+CIRCUITPY_CREATION_ID = 0x00000000
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:15]: You can go here: https://github.com/creationid/creators/blob/main/creations/heltec.md and propose a PR with a new creation ID. It doesn't matter much what it is - if there's a stock number you could use that.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    { MP_ROM_QSTR(MP_QSTR_IO17), MP_ROM_PTR(&pin_GPIO17) },
+    { MP_ROM_QSTR(MP_QSTR_IO18), MP_ROM_PTR(&pin_GPIO18) },
+    { MP_ROM_QSTR(MP_QSTR_IO19), MP_ROM_PTR(&pin_GPIO19) },
+    { MP_ROM_QSTR(MP_QSTR_IO20), MP_ROM_PTR(&pin_GPIO20) },
+    { MP_ROM_QSTR(MP_QSTR_IO21), MP_ROM_PTR(&pin_GPIO21) },
+    { MP_ROM_QSTR(MP_QSTR_IO22), MP_ROM_PTR(&pin_GPIO22) },
+    { MP_ROM_QSTR(MP_QSTR_IO23), MP_ROM_PTR(&pin_GPIO23) },
+    /* Not available on esp32:
+    { MP_ROM_QSTR(MP_QSTR_IO24), MP_ROM_PTR(&pin_GPIO24) },
+    */
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c:43]: Take out the commented-out pins.

Also only include pins that are accessible. If some are not connected at all to any accessible pads, take them away also.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    // LORA SPI, Reset, and More
+    /** Todo: Adapt for the Lora Version of the WifiKit 32 V2 board
+    { MP_ROM_QSTR(MP_QSTR_LORA_NSS), MP_ROM_PTR(&pin_GPIO8) },
+    { MP_ROM_QSTR(MP_QSTR_LORA_SCK), MP_ROM_PTR(&pin_GPIO9) },
+    { MP_ROM_QSTR(MP_QSTR_LORA_MOSI), MP_ROM_PTR(&pin_GPIO10) },
+    { MP_ROM_QSTR(MP_QSTR_LORA_MISO), MP_ROM_PTR(&pin_GPIO11) },
+    { MP_ROM_QSTR(MP_QSTR_LORA_RST), MP_ROM_PTR(&pin_GPIO12) },
+    { MP_ROM_QSTR(MP_QSTR_LORA_BUSY), MP_ROM_PTR(&pin_GPIO13) },
+    */
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c:143]: Remove from this board.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/board.c [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    0xda, 1, 0x12, // 0x12 cause 64 height, but 0x02 if it's 32 height
+    0xdb, 1, 0x40, // Set vcom configuration
+    0x8d, 1, 0x14, // Enable charge pump
+    0xAF, 0, // DISPLAY_ON
+};
+
+digitalio_digitalinout_obj_t display_on;
+
+static void display_init(void) {
+    // Need to bring GPIO4 high or the screen doesn't get power : TODO: This is GPIO21 (OLED_RST) on Heltec Lora V3 on ESP32-S3, right?
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/board.c:57]: Could you address this to-do?

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+DEBUG = 0
+ESPTOOL_FLAGS = --before=default_reset --after=hard_reset --baud 921600
+
+# This board doesn't have USB by default, it
+# instead uses a CP2102 USB-to-Serial chip
+CIRCUITPY_USB = 0
+CIRCUITPY_ESP_USB_SERIAL_JTAG = 0
+
+# Todo: qio or dio?
+CIRCUITPY_ESP_FLASH_MODE = dio
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:31]: You should be able to tell this from the module or the schematic.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+# esp-idf-config/sdkconfig-flash-26m.defaults does not exist in main branch
+CIRCUITPY_ESP_FLASH_FREQ = 26m
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:33]: Is this actually the flash frequency, or just the crystal frequency? You can choose the flash frequency.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+IDF_TARGET = esp32
+
+DEBUG = 0
+ESPTOOL_FLAGS = --before=default_reset --after=hard_reset --baud 921600
+
+# This board doesn't have USB by default, it
+# instead uses a CP2102 USB-to-Serial chip
+CIRCUITPY_USB = 0
+CIRCUITPY_ESP_USB_SERIAL_JTAG = 0
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:28]: Check the defaults for these for ESP32. I think they will default to `0`?

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+CIRCUITPY_ESPCAMERA = 0
+
+CIRCUITPY_DISPLAYIO = 1
+
+# This has something to do with XTAL Config support?:
+CIRCUITPY_CANIO = 0
+
+FROZEN_MPY_DIRS += $(TOP)/frozen/Adafruit_CircuitPython_Display_Shapes
+FROZEN_MPY_DIRS += $(TOP)/frozen/Adafruit_CircuitPython_Display_Text
+FROZEN_MPY_DIRS += $(TOP)/frozen/Adafruit_CircuitPython_RFM9x
```
Comment (dhalbert) [inline ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:47]: This board does not have an onboard RFM, does it?

Why freeze the display libraries?

**Review by dhalbert — CHANGES_REQUESTED** [review]
Hi - great progress. I noted several things.

Check some existing ESP32 boards to see what they do, especially for mpconfigboard.mk.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk (hunk 1 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:15]: Done.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c (hunk 3 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/pins.c:143]: Done.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/board.c (hunk 4 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/board.c:57]: Done.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk (hunk 5 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:31]: It has no influence to flashing process nor what debug output at startup reports: It's always mentioned "dio".

I set it to dio for now. Done.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk (hunk 7 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:28]: Exactly as you said. I removed it.

File: ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk (hunk 8 above)
Comment (szsoftware) [reply ports/espressif/boards/heltec_esp32_wifikit_32_v2/mpconfigboard.mk:47]: RFM was a leftover from the (only) other Heltec Board files I took for reference.
Removed it.

I just learn about the circuitpython framework and building process and after studying the freeze terms and ideas I wonder myself why one put this into default firmware.

However also this is a leftover from the other heltec board available in Circuitpython and there it passed the pull request/implementation process.

I remove it for now.

**Review by bill88t — CHANGES_REQUESTED** [review]

**Comment by dhalbert** [comment]
Hi @szsoftware Are you still interested in working on this? If not, we can close the PR for now, and reopen it as needed. Note that the format of the header of the board files has changed fairly recently.

**Pull request closed without merging**